File: js/compressed-token/src/v3/instructions/mint-to.ts
import {
    PublicKey,
    SystemProgram,
    TransactionInstruction,
} from '@solana/web3.js';
import { Buffer } from 'buffer';
import { CTOKEN_PROGRAM_ID } from '@lightprotocol/stateless.js';

/**
 * Parameters for creating a MintTo instruction.
 */
export interface CreateMintToInstructionParams {
    /** Mint account (CMint - decompressed compressed mint) */
    mint: PublicKey;
    /** Destination CToken account to mint to */
    destination: PublicKey;
    /** Amount of tokens to mint */
    amount: number | bigint;
    /** Mint authority (must be signer) */
    authority: PublicKey;
    /** Maximum lamports for rent and top-up combined. Transaction fails if exceeded. (0 = no limit) */
    maxTopUp?: number;
    /** Optional fee payer for rent top-ups. If not provided, authority pays. */
    feePayer?: PublicKey;
}

/**
 * Create instruction for minting tokens to a CToken account.
 *
 * This is a simple 3-4 account instruction for minting to decompressed CToken accounts.
 * Uses discriminator 7 (CTokenMintTo).
 *
 * @param params - Mint instruction parameters
 * @returns TransactionInstruction for minting tokens
 */
export function createMintToInstruction(
    params: CreateMintToInstructionParams,
): TransactionInstruction {
    const { mint, destination, amount, authority, maxTopUp, feePayer } = params;

    // Authority is writable only when maxTopUp is set AND no feePayer
    // (authority pays for top-ups only if no separate feePayer)
    const authorityWritable = maxTopUp !== undefined && !feePayer;

    const keys = [
        { pubkey: mint, isSigner: false, isWritable: true },
        { pubkey: destination, isSigner: false, isWritable: true },
        { pubkey: authority, isSigner: true, isWritable: authorityWritable },
        // System program required for rent top-up CPIs
        { pubkey: SystemProgram.programId, isSigner: false, isWritable: false },
    ];

    // Add fee_payer if provided (must be signer and writable)
    if (feePayer) {
        keys.push({ pubkey: feePayer, isSigner: true, isWritable: true });
    }

    // Build instruction data: discriminator (7) + amount (u64) + optional max_top_up (u16)
    const amountBigInt = BigInt(amount.toString());
    const dataSize = maxTopUp !== undefined ? 11 : 9; // 1 + 8 + optional 2
    const data = Buffer.alloc(dataSize);

    data.writeUInt8(7, 0); // CTokenMintTo discriminator
    data.writeBigUInt64LE(amountBigInt, 1);

    if (maxTopUp !== undefined) {
        data.writeUInt16LE(maxTopUp, 9);
    }

    return new TransactionInstruction({
        programId: CTOKEN_PROGRAM_ID,
        keys,
        data,
    });
}

File: sdk-libs/token-client/src/actions/mint_to.rs
//! Mint tokens action for Light Token.
//!
//! Simple interface for minting tokens to a Light Token account.

use light_client::rpc::{Rpc, RpcError};
use light_token::instruction::MintTo as MintToInstruction;
use solana_instruction::Instruction;
use solana_keypair::Keypair;
use solana_pubkey::Pubkey;
use solana_signature::Signature;
use solana_signer::Signer;

/// Parameters for minting tokens to a Light Token account.
///
/// # Example
/// ```ignore
/// MintTo {
///     mint,
///     destination,
///     amount: 1000,
///     ..Default::default()
/// }.execute(&mut rpc, &payer, &mint_authority).await?;
/// ```
#[derive(Default, Clone, Debug)]
pub struct MintTo {
    /// The mint public key.
    pub mint: Pubkey,
    /// The destination token account.
    pub destination: Pubkey,
    /// Amount of tokens to mint.
    pub amount: u64,
}

pub fn create_mint_to_instructions(
    mint_to: &MintTo,
    fee_payer: Pubkey,
    authority: Pubkey,
) -> Result<Vec<Instruction>, RpcError> {
    let ix = MintToInstruction {
        mint: mint_to.mint,
        destination: mint_to.destination,
        amount: mint_to.amount,
        authority,
        fee_payer,
    }
    .instruction()
    .map_err(|e| RpcError::CustomError(format!("Failed to create instruction: {}", e)))?;

    Ok(vec![ix])
}

impl MintTo {
    pub fn instructions(
        &self,
        fee_payer: Pubkey,
        authority: Pubkey,
    ) -> Result<Vec<Instruction>, RpcError> {
        create_mint_to_instructions(self, fee_payer, authority)
    }

    /// Execute the mint_to action via RPC.
    ///
    /// # Arguments
    /// * `rpc` - RPC client
    /// * `payer` - Transaction fee payer keypair (also pays for rent top-ups)
    /// * `authority` - The mint authority keypair
    ///
    /// # Returns
    /// `Result<Signature, RpcError>` - The transaction signature
    pub async fn execute<R: Rpc>(
        self,
        rpc: &mut R,
        payer: &Keypair,
        authority: &Keypair,
    ) -> Result<Signature, RpcError> {
        let instructions = create_mint_to_instructions(&self, payer.pubkey(), authority.pubkey())?;

        let mut signers: Vec<&Keypair> = vec![payer];
        if authority.pubkey() != payer.pubkey() {
            signers.push(authority);
        }

        rpc.create_and_send_transaction(&instructions, &payer.pubkey(), &signers)
            .await
    }
}
